### ISSUE TITLE
Update Datepicker UX

### ISSUE THREAD

**Issue by dannymcclain** [Expensify/App#52621]
## Problem
Currently our datepicker inputs always show the calendar popover, and this leads to a pretty cumbersome interface in places like the search filters where we have both a before and after date input.

![CleanShot 2024-11-15 at 08 54 52@2x](https://github.com/user-attachments/assets/dd5dbb93-1588-4349-bc37-de1284e8b787)

## Solution
Coming from [this thread](https://expensify.slack.com/archives/C03TME82F/p1728674000738159), let’s update our datepicker so that the calendar popover only appears when the input is focused.

https://github.com/user-attachments/assets/23ee9782-1e20-46ca-80fb-80a35706ed80

![image](https://github.com/user-attachments/assets/fded8266-02df-4d98-a662-3ad2286a6162)

#### Notes:
- For this update, we’re going to leave the placeholder text as is. Focusing the input should only change the border to green and make the popover appear. (It should not bring up the keyboard on mobile.)
- Selecting a date in the popover should set the date in the input and blur the input (thereby dismissing the calendar popover).
- We also need to update the selected state of the date to match our designs (green background, bold text, with the same text color we use for our green buttons).
- For the date filter in search and other places where there may be multiple inputs, we **do not** want to auto-focus the first date input (as this would hide the input below it), but in situations where there is only one date input (like clearing your status for example) we **do** want to autofocus the input to reveal the popover right away.
- We plan on doing a “phase 2” update to the datepicker where we will allow users to select a date by either using the popover OR by typing a date into the input, likely using some kind of guided input ([example](https://nextui.org/docs/components/date-picker)), so let’s keep that in mind as we’re making this update. But to be clear, for phase 1, we’re not planning on tackling the guided input aspect.

[Figma file](https://www.figma.com/design/dhL5qHqYtSTUEcSkHTAdx0/Datepicker-UX?node-id=1-2&node-type=canvas&t=oHpzB37ZD2gzJXsu-11)

cc @Expensify/design

**Comment by dannymcclain** [comment]
@JmillsExpensify any opinion on which project makes the most sense for this? I was thinking `#retain` but I will defer to the smart people!

**Comment by Pholenk** [comment]
Hi @dannymcclain I can't see the image and video attached. Can you reupload it, please?

**Comment by Pholenk** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Update Datepicker UX

### What is the root cause of that problem?
UX enhancement

### What changes do you think we should make in order to solve the problem?
- Add this line to track whether the input is focused, popover relative position, and container height
```
    const [isFocused, setIsFocused] = useState(false);
    const [popoverPosition, setPopoverPosition] = useState({
        top: 0,
        left: 0,
        bottom: 0,
        right: 0,
    });
    const [containerSize, setContainerSize] = useState({
        height: 50,
    });
```
to this line https://github.com/Expensify/App/blob/4bb19d77bed6cc713cc9d159f02ee587339518cd/src/components/DatePicker/index.tsx#L74

- Add this line to set reference locally so the functionality of the popover and the text input is not dependent to the ref props
```
const localRef = useRef<BaseTextInputRef>(null);
```

- Add these lines to set popover show or hide dynamically
```
    const popoverDisplayStyle = useMemo(() => {
        if (isFocused) {
            return styles.datePickerPopoverShow;
        }

        return styles.datePickerPopoverHide;
    }, [isFocused, styles.datePickerPopoverHide, styles.datePickerPopoverShow]);
```

- Change the styles in [here](https://github.com/Expensify/App/blob/4bb19d77bed6cc713cc9d159f02ee587339518cd/src/styles/index.ts#L4056)
```diff
+        datePickerRoot: {
-             position: 'relative',
-             zIndex: 99,
+            width: '100%',
+        },
+        datePickerPopoverShow: {
+            display: 'flex',
+        },
+
+        datePickerPopoverHide: {
+            display: 'none',
+        },
```

- To expose the reference of this component add these lines
```
useImperativeHandle(
        ref,
        () => {
            if (!localRef.current) {
                return {} as BaseTextInputRef;
            }
            return localRef.current;
        },
        [],
    );
```

- Add these lines to simplify the changes of isFocused state
```
const toggleFocus = useCallback(() => {
        setIsFocused((prevState) => !prevState);
    }, []);
```

- Change the component code
```diff
-        <View style={styles.datePickerRoot}>
-            <View style={[shouldUseNarrowLayout ? styles.flex2 : {}, styles.pointerEventsNone]}>
+        <View style={[styles.flex0, styles.datePickerRoot, containerSize]}>
+            <View style={[shouldUseNarrowLayout ? styles.flex2 : {}]}>
                <TextInput
-                    ref={ref}
+                   ref={localRef}
                    inputID={inputID}
                    forceActiveLabel
                    icon={Expensicons.Calendar}
                    label={label}
                    accessibilityLabel={label}
                    role={CONST.ROLE.PRESENTATION}
                    value={selectedDate}
                    placeholder={placeholder ?? translate('common.dateFormat')}
                    errorText={errorText}
                    containerStyles={containerStyles}
                    textInputContainerStyles={[styles.borderColorFocus]}
-                    inputStyle={[styles.pointerEventsNone]}
                    disabled={disabled}
-                    readOnly
+                    readOnly={false}
+                    disableKeyboard
+                    onFocus={toggleFocus}
+                    onBlur={toggleFocus}
+                    onLayout={onLayoutHandle}
                />
            </View>
            <View
-                style={[styles.datePickerPopover, styles.border]}
+                style={[styles.datePickerPopover, styles.border, popoverDisplayStyle, popoverPosition]}
                collapsable={false}
            >
                <CalendarPicker
                    minDate={minDate}
                    maxDate={maxDate}
                    value={selectedDate}
                    onSelected={onSelected}
                />
            </View>
        </View>
```

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by dannymcclain** [comment]
@thesahindia are you the official proposal reviewer for this one?

**Comment by Pholenk** [comment]
I've [updated my proposal](https://github.com/Expensify/App/issues/52621#issuecomment-2480261556) to add the video of the change result.

cc: @thesahindia

**Comment by dannymcclain** [comment]
Not overdue. I think proposals are still being reviewed.

**Comment by dubielzyk-expensify** [comment]
Danny is OOO for a few days, but yes we should increase the space to match what he has in the mocks 👍

**Comment by dannymcclain** [comment]
Not overdue. Proposals still in review.

**Comment by maddylewis** [comment]
@kadiealexander - i added the `New Feature` label to this one to get a BZ member assigned to the issue. it looks like @thesahindia is the C+ and is reviewing proposals so not a ton for you to do here until payments are due 👍

**Comment by dannymcclain** [comment]
@daledah @Pholenk Any update on the request from @shubham1206agra above?

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

♻️ Update Datepicker UX.

### What is the root cause of that problem?

Outdated Datepicker UX which needs updated.

### What changes do you think we should make in order to solve the problem?

According to the requirements from OP:

### Functional changes

Apply the following changes to our `DatePicker` component:

<details>
<summary>DIFF 1</summary>

```diff
diff --git a/src/components/DatePicker/index.tsx b/src/components/DatePicker/index.tsx
index 7e5774fd559..dfc434e2713 100644
--- a/src/components/DatePicker/index.tsx
+++ b/src/components/DatePicker/index.tsx
@@ -1,6 +1,6 @@
 import {setYear} from 'date-fns';
 import type {ForwardedRef} from 'react';
-import React, {forwardRef, useEffect, useState} from 'react';
+import React, {forwardRef, useCallback, useEffect, useImperativeHandle, useMemo, useRef, useState} from 'react';
 import {View} from 'react-native';
 import * as Expensicons from '@components/Icon/Expensicons';
 import TextInput from '@components/TextInput';
@@ -63,6 +63,7 @@ function DatePicker(
         value,
         shouldSaveDraft = false,
         formID,
+        autoFocus,
     }: DatePickerProps,
     ref: ForwardedRef<BaseTextInputRef>,
 ) {
@@ -72,10 +73,38 @@ function DatePicker(
     const [selectedDate, setSelectedDate] = useState(value || defaultValue || undefined);
     const {shouldUseNarrowLayout} = useResponsiveLayout();
 
+    const [showPicker, setShowPicker] = useState(false);
+    const textInputRef = useRef<BaseTextInputRef>(null);
+
+    const pickerDisplayStyle = useMemo(() => {
+        if (showPicker) {
+            return styles.datePickerPopoverShow;
+        }
+
+        return styles.datePickerPopoverHide;
+    }, [showPicker, styles.datePickerPopoverHide, styles.datePickerPopoverShow]);
+
+    const toggleShowPicker = useCallback(() => {
+        setShowPicker((prevShowPicker) => !prevShowPicker);
+    }, []);
+
+    useImperativeHandle(
+        ref,
+        () => {
+            if (!textInputRef.current) {
+                return {} as BaseTextInputRef;
+            }
+            return textInputRef.current;
+        },
+        [],
+    );
+
     const onSelected = (newValue: string) => {
         onTouched?.();
         onInputChange?.(newValue);
         setSelectedDate(newValue);
+        toggleShowPicker();
+        textInputRef.current?.blur();
     };
 
     useEffect(() => {
@@ -93,9 +122,9 @@ function DatePicker(
 
     return (
         <View style={styles.datePickerRoot}>
-            <View style={[shouldUseNarrowLayout ? styles.flex2 : {}, styles.pointerEventsNone]}>
+            <View style={[shouldUseNarrowLayout ? styles.flex2 : {}]}>
                 <TextInput
-                    ref={ref}
+                    ref={textInputRef}
                     inputID={inputID}
                     forceActiveLabel
                     icon={Expensicons.Calendar}
@@ -109,11 +138,14 @@ function DatePicker(
                     textInputContainerStyles={[styles.borderColorFocus]}
                     inputStyle={[styles.pointerEventsNone]}
                     disabled={disabled}
-                    readOnly
+                    disableKeyboard
+                    onFocus={toggleShowPicker}
+                    autoFocus={autoFocus}
+                    shouldDelayFocus
                 />
             </View>
             <View
-                style={[styles.datePickerPopover, styles.border]}
+                style={[styles.datePickerPopover, styles.border, pickerDisplayStyle]}
                 collapsable={false}
             >
                 <CalendarPicker
```

</details>

**Explanation for the changes**
1. We're getting `autoFocus` prop which will be passed to `DatePicker` via `InputWrapper` and we're passing it to the `TextInput` component within `DatePicker`, this is done for the optional auto focus functionality as required in OP.
2. The rest of the changes are related to the Datepicker UX update including slight change of display styles, showing the picker when focusing the input and hiding it after selection is performed.
3. We're passing `disableKeyboard`, `autoFocus` and `shouldDelayFocus` to the TextInput to fulfil the OP requirements - we need `shouldDelayFocus` for a slight delay before focusing the TextInput which is useful when using transitions especially for Native where without the delay the picker won't show consistently.

### Styling changes

Applying the following style changes as required by OP:

**Note**: These will probably be changed slightly after some input from design team.

<details>
<summary>DIFF 2</summary>

```diff
diff --git a/src/components/DatePicker/CalendarPicker/index.tsx b/src/components/DatePicker/CalendarPicker/index.tsx
index 4e1fef2d827..becdf3bb629 100644
--- a/src/components/DatePicker/CalendarPicker/index.tsx
+++ b/src/components/DatePicker/CalendarPicker/index.tsx
@@ -247,11 +247,11 @@ function CalendarPicker({
                                     <View
                                         style={[
                                             themeStyles.calendarDayContainer,
-                                            isSelected ? themeStyles.buttonDefaultBG : {},
+                                            isSelected ? themeStyles.buttonSuccess : {},
                                             !isDisabled ? StyleUtils.getButtonBackgroundColorStyle(getButtonState(hovered, pressed)) : {},
                                         ]}
                                     >
-                                        <Text style={isDisabled ? themeStyles.buttonOpacityDisabled : {}}>{day}</Text>
+                                        <Text style={[isSelected ? themeStyles.textBold : {}, isDisabled ? {...themeStyles.buttonOpacityDisabled, ...themeStyles.textBold} : {}]}>{day}</Text>
                                     </View>
                                 )}
                             </PressableWithoutFeedback>
```

</details>

and the following style updates related to the `DatePicker` functionality update:

<details>
<summary>DIFF 3</summary>

```diff
diff --git a/src/styles/index.ts b/src/styles/index.ts
index 13920783568..c8416c178cc 100644
--- a/src/styles/index.ts
+++ b/src/styles/index.ts
@@ -4073,8 +4073,15 @@ const styles = (theme: ThemeColors) =>
         },
 
         datePickerRoot: {
-            position: 'relative',
-            zIndex: 99,
+            width: '100%',
+        },
+
+        datePickerPopoverShow: {
+            display: 'flex',
+        },
+
+        datePickerPopoverHide: {
+            display: 'none',
         },
 
         datePickerPopover: {
```

</details>

##

Final change, as an example that the `autoFocus` functionality works as requested in OP, we're passing the prop in `src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx` to the `InputWrapper` component:

```diff
diff --git a/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx b/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx
index 0d400211cc4..31c606ac766 100644
--- a/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx
+++ b/src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx
@@ -69,6 +69,7 @@ function DateOfBirthPage() {
                             defaultValue={privatePersonalDetails?.dob ?? ''}
                             minDate={subYears(new Date(), CONST.DATE_BIRTH.MAX_AGE)}
                             maxDate={subYears(new Date(), CONST.DATE_BIRTH.MIN_AGE)}
+                            autoFocus
                         />
                     </FormProvider>
                 )}
```

⬆️ This will allow testing in **Settings > Profile > Private section > Date of Birth field** which will auto focus when opened.

##

✅ Here's a [test branch](https://github.com/Expensify/App/compare/main...ikevin127:Expensify:ikevin127-updateDatePickerUX) with all the changes ready to be tested.

> Please also, create a scrollable page with lots of inputs (preferably a test form). And put the DatePicker at bottom.

I already did this and the behaviour can be observed in a video below, for reference I simply duplicated the `InputWrapper` component in `src/pages/settings/Profile/PersonalDetails/DateOfBirthPage.tsx` and removed the `autoFocus` prop.

### Result videos (before / after)

<details>
<summary>iOS: Native & MacOS: Chrome (+ Input at the bottom)</summary>

| iOS: Native | iOS: Native - Input at the bottom |
| ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/d067fbda-a442-4573-8faf-b93c6c2e2ea7"> | <video src="https://github.com/user-attachments/assets/4a03f12a-8182-4d89-815d-f41eeb4ece5a"> |


https://github.com/user-attachments/assets/72160cf1-9526-407d-aa69-c7a5ba08a864

</details>

**Comment by shawnborton** [comment]
@ikevin127 in this screensho there, it looks like any input that has a green border is "active" as in, the user is focused into it. I think the idea is that if you are indeed focused into an input, the calendar should be showing. So notice on the left that even though the input appears to be focused, there is no calendar showing:
![CleanShot 2025-01-03 at 07 38 59@2x](https://github.com/user-attachments/assets/103f8275-5354-4cfc-9810-2cba00c7b233)

I'm also not sure if we want the calendar to immediately dismiss after you make a choice? What do you think @Expensify/design? I could see where we actually kind of need to close the calendar immediately after a choice is made so that you can easily navigate to any subsequent date inputs that might be below. So that's probably a good thing.

**Comment by shawnborton** [comment]
We should try to follow what Danny outlined in the solution [above](https://github.com/Expensify/App/issues/52621#issue-2662248314), where the calendar input sits on top of the content and doesn't push the content down.

**Comment by ikevin127** [comment]
My bad, I missed the part where it's not supposed to shift the input below.

> it looks like any input that has a green border is "active" as in, the user is focused into it.

That was just the active style, the input wasn't actually focused - which was easy to remove in 1 line:

https://github.com/Expensify/App/blob/81186aaf57eedc3ec285ca09f1b8cb5e07645e06/src/components/DatePicker/index.tsx#L109

##

I spent some hours on this and concluded that this is a lot more cognitively complex then I thought initially and this is because of our modals close behaviour / custom input re-focus logic we have which works slightly different between platforms (Web / Native), some things to consider for a complete solution (mine isn't even close).

Some hard to solve issues regarding this update are:
- the focused date input which is showing the date picker will lose focus and disappear when the date picker's Year select **modal** is open to select the Year, then when the modal is closed we have to re-focus the date input and show the date picker as if nothing happened (unnoticeable to the user), showing the updated selected Year in the date picker (the hardest IMO)
- when the date input is focused, the date picker must show up under the input but on top of any other inputs which can exist under the focused date input while making sure the scroll view expands when this absolute positioned date picker is displayed, in order to not overflow out of the scroll view and be unusable (quite hard)

Even with the solution that I proposed which displaces the other inputs and expands the scroll view, we're still left with the  Year modal issue / date picker input losing focus.

**I think we'd have more luck getting this UX update with somebody from the expert agencies, specifically one that worked on the modals close behaviour / custom input re-focus logic.**

**Comment by perunt** [comment]
Hey guys,
I'm from the expert agency Margelo. @maddylewis  tagged me here ([link](https://margelo.slack.com/archives/C035J5C9FAP/p1735909592333049?thread_ts=1735859398.331599&cid=C035J5C9FAP)). I can help you with it and will be able to start this week

**Comment by dannymcclain** [comment]
@perunt Awesome! I'd love to know if the desired behavior in the OP is possible/how difficult it would be to achieve. Depending on the answer to that, I think we could open to exploring other solutions that will still improve the UX and visual design.

cc @Expensify/design & @JmillsExpensify just to keep peeps in the loop.

**Comment by dannymcclain** [comment]
Any updates on this one? Also `@product-people-of-exfy`, is this less of a priority now that we're thinking about breaking out all the different date options into push rows?

**Comment by shawnborton** [comment]
Good question... I still think it's probably worth addressing this so that can still use multiple date inputs on the same page if we wanted/needed to.

**Comment by trjExpensify** [comment]
> I still think it's probably worth addressing this so that can still use multiple date inputs on the same page if we wanted/needed to.

That makes sense to me too, maybe for `Range` for example. 

>@product-people-of-exfy

new number, who dis?

**Comment by dannymcclain** [comment]
Ah yeah good call. That makes sense to me. And for the other ones that are a single date (ie Before), we'd just auto-focus the date input when you opened the screen from the push row so you'd immediately get the calendar anyways. 👍

**Comment by shawnborton** [comment]
Sounds good to me, let's do it 👍

**Comment by dannymcclain** [comment]
Any updates here? Sounds like we still want to pursue this behavior if possible.

**Comment by perunt** [comment]
> For the date filter in search and other places where there may be multiple inputs, we do not want to auto-focus the first date input (as this would hide the input below it), but in situations where there is only one date input (like clearing your status for example) we do want to autofocus the input to reveal the popover right away.

@dannymcclain Do you want to expand this even if we have other content on the page? If yes, it may also hide some information which could be confusing.

Sorry for the delay. Currently, I'm working on a feature where when you press on a message, while the context menu is opening, we move that message to the top so it remains readable. I can apply similar techniques from that PR here as well, but maybe in the next iteration as I would like to create an example to show how it would look.

For now, if you want to have overflow behavior (position: absolute), we need to handle some edge cases, but it shouldn't take long. I haven't seen the calendar component yet, but it shouldn't be an issue to customize it as well.
Just to clarify - do you want the same fade-in effect when it appears on the screen? Maybe we already have a similar effect in the app? I want to keep it consistent across the application

**Comment by dannymcclain** [comment]
>Do you want to expand this even if we have other content on the page? If yes, it may also hide some information which could be confusing.

I think that if the field is focused, yes, the calendar popover would cover other content on the screen. So I think we need some way to decide if we want to autofocus the field or not (we might already have a mechanism for this?). That way, when there's only the date input on the screen, we could decide to autofocus the field (and automatically show the popover). But if there _is_ other info on the screen along with the date picker, we could decide _not_ to.

>Just to clarify - do you want the same fade-in effect when it appears on the screen? Maybe we already have a similar effect in the app? I want to keep it consistent across the application

I personally am not sure I think the calendar popover even needs any kind of animation. I think it could just appear similar to how our context menus just appear when you right click on a message. Open to other opinions from the @Expensify/design team on that though.

**Comment by dubielzyk-expensify** [comment]
> That way, when there's only the date input on the screen, we could decide to autofocus the field (and automatically show the popover). But if there is other info on the screen along with the date picker, we could decide not to.

I like that logic.

> I personally am not sure I think the calendar popover even needs any kind of animation. I think it could just appear similar to how our context menus just appear when you right click on a message. Open to other opinions from the @Expensify/design team on that though.

Feels like something that would be worth playing with once we have a branch to play with. I'd say a simple fade from 0 -> 1 opacity could be a good start though.

**Comment by shawnborton** [comment]
Agree about not auto-focusing if the calendar might obscure valuable information/UI behind it. 

And yeah, a simple fade or something might be nice - but that can be a follow up once we have the correct behavior in place.

**Comment by dannymcclain** [comment]
Sounds good to me!

**Comment by dannymcclain** [comment]
bump @perunt

**Comment by perunt** [comment]
I just finished with the release blocker, and this task is my number one priority right now. I'll start working on it first thing in the morning. Sorry for the delays, I will do the first examples tommorrow

**Comment by perunt** [comment]
Quick update:  I've added an expandable component and am currently working on handling its behavior properly, especially ensuring the overflow works as intended.
<Image height=700 src="https://github.com/user-attachments/assets/a9897847-af57-412b-b47e-cd356df8b784" />

Sometimes this input might be right next to the ‘Save’ button—do you want it to overlap the button, or should it push the button down instead?

Just a thought: having content appear on top of everything feels more like a web thing—I don’t really see it much on mobile. What do you think about splitting the behavior? On mobile, the expandable list could push the content below it, and on web, we stick with the overflow style.
(If it pushes content down, the PR is almost ready. If it needs to overlap or combine these approaches, I’ll need a few more hours

**Comment by shawnborton** [comment]
> do you want it to overlap the button, or should it push the button down instead?

Can we launch the calendar above the input in that case?

**Comment by shawnborton** [comment]
Oh actually that's not a terrible idea at all... I would be down to see something like that.

**Comment by dannymcclain** [comment]
>Maybe a crazy idea. On mobile, use bottom docked modal instead of normal popover similar to what we do for emoji picker.

Yeah I would be down for seeing something like this too! Seems interesting.

>On mobile, the expandable list could push the content below it, and on web, we stick with the overlap style.

Honestly, I don't think I'd be against seeing this where the content gets pushed down on all platforms. I don't love the idea of splitting the behavior, but I don't know that I'm against using the push down pattern for everything if we think that works out better. I think for the most part, these inputs will usually be on their own—we definitely need a solution for when they're not, but I don't imagine it will be the "all the time" case.

But I think using the bottom docked modal for mobile and the regular popover for desktop is more appealing to me—especially after making that connection with that same behavior in the emoji picker.

**Comment by dubielzyk-expensify** [comment]
Lots of apps do the bottom sheet so I'm totally keen to see that as well. 

> On mobile, the expandable list could push the content below it, and on web, we stick with the overlap style.

I kinda think that we should do the overlap style everywhere. I think as long as the calendar dismisses on setting the date or dismissing (tapping outside), I don't think it'll be a bad experience. As long as it's all in view that is.

**Comment by perunt** [comment]
This is just for testing purposes, so please disregard the missing functionality.

https://github.com/user-attachments/assets/eadc1364-b1bc-477d-b106-681a592916d4

https://github.com/user-attachments/assets/8f7e11d2-b4f7-4aa0-8674-879975d19fad

The main issue I faced here is selecting a year. Previously, this was done by navigating to a separate screen. Now, it seems it should appear in a popup (on the web) or in a bottom sheet modal(mobile).

**Comment by perunt** [comment]
So, the question here is: how should the year selection page look? Should it remain the same as before?
Before:

https://github.com/user-attachments/assets/dad83c9a-5b38-474c-83a3-f830a28df78a

**Comment by shawnborton** [comment]
> The main issue I faced here is selecting a year. Previously, this was done by navigating to a separate screen. Now, it seems it should appear in a popup (on the web) or in a bottom sheet modal(mobile).

Can we see what it's like to navigate from the bottom-sheet date picker, to the RHP year selection? Maybe it won't feel so terrible?

**Comment by dannymcclain** [comment]
>Can we see what it's like to navigate from the bottom-sheet date picker, to the RHP year selection? Maybe it won't feel so terrible?

Agree it'd be nice to see. I think I'm digging the bottom sheet for the calendar, so if the RHP for year selection doesn't work, I wonder if we could just push to another view inside the bottom sheet/popover and do year selection like most other date pickers? Thinking something like this:

![Image](https://github.com/user-attachments/assets/b0a12506-a64b-46ed-a58c-edd84cff1dd2)

@Expensify/design for thoughts on that!

I think it'd also be nice to adjust the styling of the calendar popover in the bottom sheet so we don't have that extra border/container inside the popover if possible.

| Current | Slight Adjustment |
|--------|--------|
| ![Image](https://github.com/user-attachments/assets/5605ea44-13c1-4219-aab9-bca79f6337c4) | ![Image](https://github.com/user-attachments/assets/458fde24-360d-4388-9ab9-75a2459adba4) |

**Comment by shawnborton** [comment]
I definitely agree with your suggestion about the style adjustment of the calendar.

As for the year, definitely curious to see the RHP prototype first. I like what you have but it does feel a little less usable for people that just want to type in the year?

**Comment by dannymcclain** [comment]
>As for the year, definitely curious to see the RHP prototype first. I like what you have but it does feel a little less usable for people that just want to type in the year?

Totally agree—just wanted to offer it as an alternate.

**Comment by shawnborton** [comment]
Hmm how long until that stuff is merged? I feel like we should try to reuse the RHP because it already exists and works great, we wouldn't need to build anything new here.

**Comment by perunt** [comment]
> Hmm how long until that stuff is merged?

Work on implementing Navigation modal screens starts Monday (https://github.com/Expensify/App/issues/53493)

> As for the year, definitely curious to see the RHP prototype first.

Here's a rough proof-of-concept using modal-in-modal that we can use temporarily:

https://github.com/user-attachments/assets/87b5842a-86fb-41c1-a3b4-f116d203b9d3

Quick demo of potential web transition (very rough, just to illustrate the concept):

https://github.com/user-attachments/assets/978cc48b-325a-43e8-ad42-8b41aa23d808

**Comment by shawnborton** [comment]
I don't mind that! On web, when you select a year, I would expect the datepicker to still be open when you get back to the original page.

**Comment by dannymcclain** [comment]
>I don't mind that!

Yeah it's really not bad! Feels way less weird than I was expecting.

>On web, when you select a year, I would expect the datepicker to still be open when you get back to the original page.

I agree. Since this video is just a rough mock, I'm assuming/hoping we could make that happen in the actual version.

**Comment by perunt** [comment]
Yeah, it should stay open until we select a date or click outside the modal. Okay, then it looks like we'll stick with this implementation. 

I'll use this design for the modal
![Image](https://github.com/user-attachments/assets/6cdef3ac-824f-4437-bd11-8c9c1d7ceddc)

I'll keep you updated on how it's going

**Comment by shawnborton** [comment]
Sounds great, I like where this is going - thanks @perunt !

**Comment by perunt** [comment]
A small update on this.

https://github.com/user-attachments/assets/c5f86cda-6176-40f4-9172-a8971f8b3286

https://github.com/user-attachments/assets/e9b91f4e-9fec-416c-b96f-bc83e2b25247

**Comment by shawnborton** [comment]
Nice! I dig it.

**Comment by dubielzyk-expensify** [comment]
Very exciting to see. Great job so far @perunt

**Comment by chrispader** [comment]
> Currently, navigating from the bottom-sheet date picker to the year selection in RHP requires implementing modal screens through Navigation, not React Native (our current approach) [@chrispader](https://github.com/chrispader) is working on adding Modal Screens for Navigation. Once complete, we'll be able to navigate from BottomSheet/modal screens to anywhere in the app.

> Hmm how long until that stuff is merged? I feel like we should try to reuse the RHP because it already exists and works great, we wouldn't need to build anything new here.

Sorry for the delay on the reply. The modal screen PR will need some more time, so i don't think waiting on the modal screen migration would be a good idea as of right now, but i see you already went that way anyway :)

**Comment by dannymcclain** [comment]
@perunt what are your thoughts on the comment above? Are we blocked from implementing this in the way you showed in your video for now?

**Comment by perunt** [comment]
No, it's not a blocker at all. In fact, I've already implemented it

I noticed a reviewer’s comment suggesting that we should use this component everywhere. Should I update the other screens as well?

**Comment by shawnborton** [comment]
I think that would be lovely.

**Comment by perunt** [comment]
There were around 10 places with the calendar component. I just replaced it and removed the old component. Testing throughout the app, and it looks good. Waiting for a review, as all the places need to be checked

**Comment by dannymcclain** [comment]
@shubham1206agra @perunt What are the next steps here? How can we push this forward this week?

**Comment by perunt** [comment]
just reverted to the old changes that blocked the merge. It should be good now

**Comment by dannymcclain** [comment]
PR is currently in progress—we're waiting on some updates from Taras (who is currently OOO). Should be able to dive back into it next week!

**Comment by grgia** [comment]
Noice!

**Comment by dannymcclain** [comment]
We are actively working through some minor issues on the PR—but this is feeling good and getting close!

**Comment by dannymcclain** [comment]
Still working through the last few little issues on the PR.

**Comment by kadiealexander** [comment]
Reassigning to handle payment as I'm heading OOO. Thanks for your help @trjExpensify!

**Comment by trjExpensify** [comment]
👋 @shubham1206agra can we get the checklist here please, and confirm the number of regressions? Sounds like just the 1 [here](https://github.com/Expensify/App/issues/60243), as the other here was [observed in the PR for a follow-up](https://github.com/Expensify/App/issues/60241#issuecomment-2805490355).

**Comment by trjExpensify** [comment]
Hm, but the [RCA](https://github.com/Expensify/App/issues/60243#issuecomment-2804150582) places that original PR at fault (CC: @daledah @mollfpr) of introducing it. If that PR wasn't the cause, what was?

**Comment by trjExpensify** [comment]
Yeah, that line reads to me "this bug only appeared because of the offending PR".

**Comment by trjExpensify** [comment]
>So the culprit here could be https://github.com/Expensify/App/pull/40757, as they created a way to "modify input values without focusing on the field" by adding clear button and overlooked this bug.

Hm, but the `X` is only shown when the the field is focused here: 

<img width="921" alt="Image" src="https://github.com/user-attachments/assets/def8cbea-4df1-456a-bbf6-7f69d49b0e29" />

Whereas with the DatePicker field it's displayed when not focused, which could be incorrect in itself? CC: @Expensify/design 

<img width="1053" alt="Image" src="https://github.com/user-attachments/assets/77df226a-e2c3-403d-8dad-72f79bf5cbd8" />

**Comment by trjExpensify** [comment]
>I've misused the word "focusing" here, what I meant was "user press/click to focus on the field"

Ah right, that makes _way_ more sense. To rephrase: 

>So the culprit here could be https://github.com/Expensify/App/pull/40757, as they created a way to modify input values when a field is focused by adding the clear button (`X`) and overlooked this bug.

**Comment by trjExpensify** [comment]
> Test Case 2: Auto-Opening Date Picker in Settings Flow
> 
> 1. Open the Settings screen.
> 2. Tap the Status button.
> 3. Tap the Clean After button.
> 4. Select Custom.
> 5. Tap the Date input field.
> 6. The DatePicker should automatically open upon focusing the date input field (since it’s the only element on the screen).
> 7. Tap Clear and verify the input field is cleared.

Modifying this one. Typo in step 3, and step 5 is unclear you need to select the `Date` push row first to get to the date input screen. 

> Test Case 2: Auto-Opening Date Picker in Settings Flow
> 
> 1. Open the Settings screen.
> 2. Tap the Status button.
> 3. Tap the Clear After button.
> 4. Select Custom.
> 5. Click the Date push row
> 6. Verify the Date input field is focused automatically (since it's the only element on the screen) and the DatePicker is revealed
> 7. Tap Clear and verify the input field is cleared.

**Comment by trjExpensify** [comment]
Paid, closing!
